File: lib/vnet/diag/diag_test.go
// Teleport
// Copyright (C) 2025 Gravitational, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

package diag

import (
	"context"
	"os/exec"
	"testing"

	"github.com/gravitational/trace"
	"github.com/jonboulle/clockwork"
	"github.com/stretchr/testify/require"

	diagv1 "github.com/gravitational/teleport/gen/proto/go/teleport/lib/vnet/diag/v1"
)

func TestGenerateReport_FailedCheck(t *testing.T) {
	diagCheck := &FakeDiagCheck{shouldFail: true}

	report, err := GenerateReport(context.Background(), ReportPrerequisites{
		Clock:               clockwork.NewFakeClock(),
		NetworkStackAttempt: &diagv1.NetworkStackAttempt{},
		DiagChecks:          []DiagCheck{diagCheck},
	})

	require.NoError(t, err)
	require.Len(t, report.Checks, 1)
	checkAttempt := report.Checks[0]
	require.Equal(t, diagv1.CheckAttemptStatus_CHECK_ATTEMPT_STATUS_ERROR, checkAttempt.Status)

	// Verify that commands are still included even if the check itself failed.
	require.Len(t, checkAttempt.Commands, 1)
	// Verify that CheckReport is not empty, as otherwise it'd be impossible to tell the kind of the
	// check that failed.
	require.NotNil(t, checkAttempt.CheckReport)
}

func TestGenerateReport_SkipCommands(t *testing.T) {
	diagCheck := &FakeDiagCheck{}

	report, err := GenerateReport(context.Background(), ReportPrerequisites{
		Clock:               clockwork.NewFakeClock(),
		NetworkStackAttempt: &diagv1.NetworkStackAttempt{},
		DiagChecks:          []DiagCheck{diagCheck},
		SkipCommands:        true,
	})

	require.NoError(t, err)
	require.Len(t, report.Checks, 1)
	checkAttempt := report.Checks[0]
	require.Equal(t, diagv1.CheckAttemptStatus_CHECK_ATTEMPT_STATUS_OK, checkAttempt.Status)
	require.Empty(t, checkAttempt.Commands)
}

type FakeDiagCheck struct {
	shouldFail bool
}

func (f *FakeDiagCheck) Run(ctx context.Context) (*diagv1.CheckReport, error) {
	if f.shouldFail {
		return nil, trace.Errorf("something went wrong")
	}

	return &diagv1.CheckReport{
		Report: &diagv1.CheckReport_RouteConflictReport{
			RouteConflictReport: &diagv1.RouteConflictReport{
				RouteConflicts: []*diagv1.RouteConflict{},
			},
		},
	}, nil
}

func (f *FakeDiagCheck) Commands(ctx context.Context) []*exec.Cmd {
	return []*exec.Cmd{exec.CommandContext(ctx, "echo", "foo")}
}

func (f *FakeDiagCheck) EmptyCheckReport() *diagv1.CheckReport {
	return &diagv1.CheckReport{}
}

File: web/packages/teleterm/src/services/vnet/diag.test.ts
/**
 * Teleport
 * Copyright (C) 2025 Gravitational, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import * as diag from 'gen-proto-ts/teleport/lib/vnet/diag/v1/diag_pb';

import { reportToText } from './diag';
import {
  makeCheckAttempt,
  makeCheckReport,
  makeCommandAttempt,
  makeDNSReport,
  makeDNSZoneResult,
  makeRecordResult,
  makeReport,
  makeRouteConflict,
  makeVNetDNSReachability,
} from './testHelpers';

describe('reportToText', () => {
  it('converts report correctly', () => {
    const routeConflictReport = makeCheckReport({
      status: diag.CheckReportStatus.ISSUES_FOUND,
    });
    routeConflictReport.report = {
      oneofKind: 'routeConflictReport',
      routeConflictReport: {
        routeConflicts: [
          makeRouteConflict({
            dest: '100.64.0.0/10',
            vnetDest: '100.64.0.1',
            interfaceName: 'utun5',
            interfaceApp: 'VPN: Foobar',
          }),
          makeRouteConflict({
            dest: '0.0.0.0/1',
            vnetDest: '100.64.0.0/10',
            interfaceName: 'utun6',
            interfaceApp: '',
          }),
        ],
      },
    };
    const dnsCheckReport = makeCheckReport({
      status: diag.CheckReportStatus.ISSUES_FOUND,
      report: {
        oneofKind: 'dnsReport',
        dnsReport: makeDNSReport({
          ipv6Reachability: makeVNetDNSReachability({
            address: '[fdff:fd74:46c0::2]:53',
            reachable: false,
            respondedA: false,
            respondedAaaa: false,
            error:
              'dial udp [fdff:fd74:46c0::2]:53: connect: network unreachable',
          }),
          zoneResults: [
            makeDNSZoneResult({ zone: 'company.test' }),
            makeDNSZoneResult({
              zone: 'hijack.zone.test',
              aRecord: makeRecordResult({
                status: diag.DNSZoneStatus.DNS_ZONE_STATUS_HIJACKED,
                observedIp: '1.2.3.4',
              }),
              aaaaRecord: makeRecordResult({
                status: diag.DNSZoneStatus.DNS_ZONE_STATUS_HIJACKED,
                observedIp: '2001:db8::1234',
              }),
            }),
            makeDNSZoneResult({
              zone: 'mixed.zone.test',
              aRecord: makeRecordResult(),
              aaaaRecord: makeRecordResult({
                status: diag.DNSZoneStatus.DNS_ZONE_STATUS_HIJACKED,
                observedIp: '2001:db8::1234',
              }),
            }),
            makeDNSZoneResult({
              zone: 'not-registered.zone.test',
              aRecord: makeRecordResult({
                status: diag.DNSZoneStatus.DNS_ZONE_STATUS_NOT_REGISTERED,
              }),
              aaaaRecord: makeRecordResult({
                status: diag.DNSZoneStatus.DNS_ZONE_STATUS_NOT_REGISTERED,
              }),
            }),
            makeDNSZoneResult({
              zone: 'staging.test',
              aRecord: makeRecordResult({
                status: diag.DNSZoneStatus.DNS_ZONE_STATUS_TIMEOUT,
                error: 'i/o timeout',
              }),
              aaaaRecord: makeRecordResult({
                status: diag.DNSZoneStatus.DNS_ZONE_STATUS_TIMEOUT,
                error: 'i/o timeout',
              }),
            }),
          ],
        }),
      },
    });
    const sshConfigReport = makeCheckReport({
      status: diag.CheckReportStatus.OK,
    });
    sshConfigReport.report = {
      oneofKind: 'sshConfigurationReport',
      sshConfigurationReport: {
        userOpensshConfigPath: '~/.ssh/config',
        vnetSshConfigPath:
          '/Users/user/Library/Application Support/Teleport Connect/tsh/vnet_ssh_config',
        userOpensshConfigIncludesVnetSshConfig: false,
        userOpensshConfigExists: false,
        userOpensshConfigContents: '',
      },
    };
    const report = makeReport({
      checks: [
        makeCheckAttempt({
          checkReport: routeConflictReport,
          commands: [makeCommandAttempt()],
        }),
        makeCheckAttempt({
          checkReport: dnsCheckReport,
        }),
        makeCheckAttempt({
          checkReport: sshConfigReport,
        }),
      ],
    });

    const actualText = reportToText(report);
    expect(actualText).toMatchSnapshot();
    // Verify that the text ends with a newline.
    expect(actualText.endsWith('\n')).toBe(true);
  });

  it('renders DNS report when both nameservers are unreachable', () => {
    const dnsCheckReport = makeCheckReport({
      status: diag.CheckReportStatus.ISSUES_FOUND,
      report: {
        oneofKind: 'dnsReport',
        dnsReport: makeDNSReport({
          ipv4Reachability: makeVNetDNSReachability({
            address: '100.64.0.2:53',
            reachable: false,
            respondedA: false,
            respondedAaaa: false,
            error: 'dial udp 100.64.0.2:53: connect: connection refused',
          }),
          ipv6Reachability: makeVNetDNSReachability({
            address: '[fdff:fd74:46c0::2]:53',
            reachable: false,
            respondedA: false,
            respondedAaaa: false,
            error:
              'dial udp [fdff:fd74:46c0::2]:53: connect: connection refused',
          }),
          zoneResults: [],
        }),
      },
    });
    const report = makeReport({
      checks: [makeCheckAttempt({ checkReport: dnsCheckReport })],
    });

    const actualText = reportToText(report);
    expect(actualText).toMatchSnapshot();
    expect(actualText.endsWith('\n')).toBe(true);
  });
});
